### ISSUE TITLE
Running out of connections during INSERT .. SELECT with repartitioning

### ISSUE THREAD

**Issue by JelteF** [citusdata/citus#3439]
During tests we sometimes run out of connections now. This seems to be related to the new INSERT .. SELECT code introduced in #3376, since it is one of the tests that fails and tests that run in parallel with it fail as well. 

Example of a failing test: https://circleci.com/gh/citusdata/citus/84922?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

This should be investigated before 9.2 release, to ensure that we are not leaking connections or opening too many connections. If that's not the case a simple solution to the flaky test would probably be not running it in parallel with other tests.

**Comment by onderkalaci** [comment]
```
parallel group (6 tests):  window_functions multi_insert_select_window recursive_dml_with_different_planners_executors dml_recursive multi_shard_update_delete insert_select_repartition
```

`window_functions` operates on 32 shard table
`multi_insert_select_window` operates on 32 shard table
`recursive_dml_with_different_planners_executors` operates on 4 shard table
`dml_recursive`  operates on 4 shard table
`multi_shard_update_delete`  operates on 6 shard table

Total: except for `insert_select_repartition`,  max 78 connections are opened concurrently, each worker 39 connections. Also, few additional connections for deadlock detection/2PC recovery, intermediate results etc.

So, it feels like `insert_select_repartition` might be leaking some connections considering the default max_connection, which  is 100.

**Comment by onderkalaci** [comment]
> Total: except for insert_select_repartition, max 78 connections are opened concurrently, each worker 39 connections. Also, few additional connections for deadlock detection/2PC recovery, intermediate results etc.

correction: there are  also few other repartition joins in two of the files, so the calculation might not be fully accurate.

**Comment by pykello** [comment]
I am still looking into this. No conclusion or any sign of connection leakage yet.

recursive_dml_with_different_planners_executors and dml_recursive seem to use around max of 20 connections per worker.

I added some debugging info to those sql files in https://github.com/citusdata/citus/pull/3442 to count number of active connections after some of the statements. I also made them sequential so we can see their connection counts individually.

You can see the failed results of check-multi there to see the connection counts. Summary is that I see the following max connections at a time per worker for those tests:

dml_recursive: 20
recursive_dml_with_different_planners_executors: 21
insert_select_repartition: 7
window_functions: 5
multi_shard_update_delete: 5
multi_insert_select_window: 6

Which add up to about 65, which is still much less than 100.

I'll continue looking into this tomorrow.

**Comment by pykello** [comment]
I found an alternative way to count connections used by each of these tests which is explained in #3448. Now I see that it can get to up to 90 connections per worker, which is too close to the 100 limit. Depending on how adapative executor slow start algorithm and some other variables behave, this can reach the 100.

I opened a PR #3448 to break this parallel group into 2 and also add a test to verify that insert/select with repartitioning isn't leaking connections.
